src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalSessionSocket.java

```diff
@@ -309,15 +326,42 @@ public void dispatchInput(int inputSequence,
       default:
          break;
       }
-      
    }
    
    /**
     * Send output to the terminal emulator.
     * @param output text to send to the terminal
+    * @param detectLocalEcho local-echo detection
     */
-   public void dispatchOutput(String output)
+   public void dispatchOutput(String output, boolean detectLocalEcho)
    {
+      if (detectLocalEcho && !localEcho_.isEmpty())
+      {
+         String lastOutput = "";
+         while (!localEcho_.isEmpty() && lastOutput.length() < output.length())
```

Comment:
This handles the case where the output received from the server includes some or all of the local output. Is it possible that the output received from the server includes all of the local output *and some additional text*? or is that handled elsewhere? (e.g. `set -o vi` for a pathological case)? If so, it seems like this should also have a mode where it emits the characters from `output` that were not consumed by `localEcho_`.